feat: reprovider

vasco-santos opened the pull request:
In the context of `js-ipfs` reannouncing blocks to the network [ipfs/js-ipfs#2160](https://github.com/ipfs/js-ipfs/issues/2160), this PR aims to add a Provider layer controlled by `js-ipfs`. This provider will be used by bitswap.

Needs:

- [ ] Different strategies
- [ ] [ipfs/js-ipfs-bitswap#199](https://github.com/ipfs/js-ipfs-bitswap/pull/199)
- [ ] documentation - should be a part of [ipfs/js-ipfs#2204](https://github.com/ipfs/js-ipfs/issues/2204)

vasco-santos committed:
feat: reprovider

vasco-santos commented:
@alanshaw @dirkmc could I get your feedback on this PR?

It is important to also look at the `bitswap`PR in this context.

FYI, for getting this to the finish line I only need to add the proper logic for the reprovide strategies. However, there is some code in Dirk's PR for the garbage collection [ipfs/js-ipfs#2022](https://github.com/ipfs/js-ipfs/pull/2022/) that would be great to use for this. So, it would be great to have that merged first.

dirkmc commented on src/core/provider/index.js:
```diff
@@ -0,0 +1,93 @@
+   * It keeps track of which blocks are provided, and allow them to be reprovided
+   * @param {Libp2p} libp2p
+   * @param {Blockstore} blockstore
+   * @param {object} options
```
Could we add a description of each option here

dirkmc commented on src/core/provider/index.js:
```diff
@@ -0,0 +1,93 @@
+    // handle options
+    const strategy = this._options.strategy || 'all'
+    const humanInterval = this._options.Interval || '12h'
+    const interval = await promisify((callback) => human(humanInterval, callback))()
```
It's a shame this code is complicated by human-to-milliseconds requiring a callback. It seems from [the source code](https://github.com/hacdias/human-to-milliseconds/blob/master/src/index.js) that it doesn't need to be a callback. If we can replace this with something synchronous then you don't need to worry about bitswap starting before provider in start.js above

dirkmc commented on src/core/provider/index.js:
```diff
@@ -0,0 +1,93 @@
+  }
+
+  /**
+   * Announce block to the network and add and entry to the tracker
```
```suggestion
   * Announce block to the network and add an entry to the tracker
```

dirkmc commented on src/core/provider/index.js:
```diff
@@ -0,0 +1,93 @@
+    })()
+  }
+
+  async findProviders (cid, options) { // eslint-disable-line require-await
```
We should probably document the options here (even though they're the same as content routing `findProviders()`

dirkmc commented on src/core/provider/queue.js:
```diff
@@ -0,0 +1,110 @@
+const log = debug('ipfs:provider')
+log.error = debug('ipfs:provider:error')
+
+class WorkerQueue {
```
Can you use p-queue instead? It will take care of the Promise execution for you

dirkmc commented on src/core/provider/reprovider.js:
```diff
@@ -0,0 +1,89 @@
+const { blockKeyToCid } = require('../utils')
+
+// const initialDelay = 15000
+const initialDelay = 3000
```
Let's add this to config

dirkmc commented on src/core/provider/reprovider.js:
```diff
@@ -0,0 +1,89 @@
+   * Reprovider goal is to reannounce blocks to the network.
+   * @param {object} contentRouting
+   * @param {Blockstore} blockstore
+   * @param {object} options
```
Let's document the options

dirkmc commented on src/core/provider/reprovider.js:
```diff
@@ -0,0 +1,89 @@
+      } else if (this._options.strategy === 'pinned') {
+
+      }
+
```
```suggestion
if (!this._timeoutId) return
```

dirkmc commented on src/core/provider/reprovider.js:
```diff
@@ -0,0 +1,89 @@
+      }
+
+      await this._worker.execute(blocks)
+
```
```suggestion
if (!this._timeoutId) return
```

dirkmc commented on src/core/provider/reprovider.js:
```diff
@@ -0,0 +1,89 @@
+      await this._worker.execute(blocks)
+
+      // Each subsequent walk should run on a `this._options.interval` interval
+      await new Promise(resolve => {
```
I'd suggest using a timeout which recursively calls `_runPeriodically()`
That way you can cancel the timeout when the reprovider is stopped

dirkmc commented:
Looks like it's on the right track @vasco-santos 👍 
I left a few comments for minor things in the code.

I wonder if it would make sense to persist the time at which the last provide occurred, so that if a node restarts it doesn't wait the full 12 hours to reprovide?

vasco-santos committed:
chore: address dirk review

vasco-santos commented:
Thanks @dirkmc ! I think that I addressed all your comments!

Regarding persisting the time of the last provide, the first reprovide will happen once the node starts, after a short delay. So, I think it should not be necessary

dirkmc commented on src/core/provider/index.js:
```diff
@@ -0,0 +1,106 @@
+    this._running = true
+
+    // handle options (config uses uppercase)
+    const humanDelay = this._options.Delay || this._options.delay || '15s'
```
Is `0` a valid value for the delay?

vasco-santos replied:
Ok, so I think we should not accept 0, as we should wait until we are connected to some peers. Do you agree to not allow 0?

As we are using a string, that validation is very annoying 😢

dirkmc replied:
Yes I think that makes sense. The validation is a little tricky 🤔

dirkmc commented on src/core/components/start.js:
```diff
@@ -56,9 +60,15 @@ module.exports = (self) => {
 
+        if (!get(self._options, 'offline') &&
+          (get(self._options, 'libp2p.config.dht.enabled', false) || get(self._options, 'libp2p.modules.contentRouting', false))) {
+          self._provider.start()
```
Because `Provider.start()` is async, bitswap may start before provider finishes starting. It may be safer to wait till `Provider.start()` has completed before starting bitswap

vasco-santos replied:
I switched the provider start to the previous function of the `series` array, and "callbackifyed" it to guarantee we wait for it

dirkmc replied:
The logic makes sense. I would suggest moving all the new code to another entry in the `series` array

vasco-santos replied:
I can't move the Provider instantiation, as I would have to get the config again, or keeping it in a temporary constant. I would prefer to keep it there, and move the starting code to the new entry of the series array

dirkmc replied:
Ok makes sense

dirkmc commented on src/core/provider/index.js:
```diff
@@ -0,0 +1,106 @@
+   * @returns {void}
+   */
+  stop () {
+    this._running = true
```
```suggestion
    this._running = false
```

dirkmc commented on src/core/provider/reprovider.js:
```diff
@@ -0,0 +1,95 @@
+   * @param {object} contentRouting
+   * @param {Blockstore} blockstore
+   * @param {object} options
+   * @param {string} options.delay reprovider initial delay in human friendly time
```
```suggestion
   * @param {string} options.delay reprovider initial delay in milliseconds
```

dirkmc commented on src/core/provider/reprovider.js:
```diff
@@ -0,0 +1,95 @@
+   * @param {Blockstore} blockstore
+   * @param {object} options
+   * @param {string} options.delay reprovider initial delay in human friendly time
+   * @param {string} options.interval reprovider interval in human friendly time
```
```suggestion
   * @param {string} options.interval reprovider interval in milliseconds
```

dirkmc commented on src/core/provider/reprovider.js:
```diff
@@ -0,0 +1,95 @@
+  async _provideBlock (block) {
+    const cid = blockKeyToCid(block.key.toBuffer())
+
+    await promisify((callback) => {
```
Is there a way we can abort this Promise if the Reprovider is stopped?

vasco-santos replied:
Once we do the DHT refactor, if we add the abortable signal to provide there, as well as to `js-libp2p` and to `delegated-content-routing`, we should be able to do that

vasco-santos committed:
chore: apply suggestions from code review

Co-Authored-By: dirkmc <dirkmdev@gmail.com>

vasco-santos committed:
chore: use new human-to-milliseconds

alanshaw commented:
@vasco-santos what's the status of this PR - still draft?

vasco-santos commented:
I was waiting on [https://github.com/ipfs/js-ipfs/pull/2022](https://github.com/ipfs/js-ipfs/pull/2022) getting merged, which had a lot of useful code to integrate here without duplicating.

daviddias commented:
Hi! js-ipfs master just got a whole new set of automated tests with #2528, https://github.com/ipfs/js-ipfs/pull/2440 and also running some of the test suites from our early testers (hi5 to  @achingbrain for setting it all up!). **Would you mind rebasing the master branch on this PR to ensure it runs all the latest tests? Thank you!**

achingbrain commented:
I'm going to close this because I think since the `.provide` method lives on `libp2p`'s `ContentRouter` interface, re-provides should be handled by the various implementations of content routers since they all may have different rules about provider record validity, etc.

Some may call out to a remote service that does the re-provide for us, for example, or some may have shorter lived records than others, who knows!
